src/librustc_mir/transform/promote_consts.rs

```diff
@@ -200,14 +203,30 @@ impl<'tcx> Visitor<'tcx> for Collector<'_, 'tcx> {
     fn visit_rvalue(&mut self, rvalue: &Rvalue<'tcx>, location: Location) {
         self.super_rvalue(rvalue, location);
 
-        match *rvalue {
+        match rvalue {
             Rvalue::Ref(..) => {
                 self.candidates.push(Candidate::Ref(location));
             }
-            Rvalue::Repeat(..) if self.tcx.features().const_in_array_repeat_expressions => {
-                // FIXME(#49147) only promote the element when it isn't `Copy`
-                // (so that code that can copy it at runtime is unaffected).
-                self.candidates.push(Candidate::Repeat(location));
+            Rvalue::Repeat(elem, n)
+                if self.ccx.tcx.features().const_in_array_repeat_expressions =>
+            {
+                let is_copy = elem
+                    .ty(&self.ccx.body.local_decls, self.ccx.tcx)
+                    .is_copy_modulo_regions(self.ccx.tcx, self.ccx.param_env, self.span);
+                let n = n.try_eval_usize(self.ccx.tcx, self.ccx.param_env);
+                let length_requires_copy_or_promotion = match n {
+                    // Unevaluable (e.g. too generic) -> assume > 1.
+                    None => true,
```

Comment:
So could this lead to a problem if we later figure out that actually, the constant *is* `1`, and then we compile the program after a failed attempt to explicitly promote? (That would trigger the delayed_span_bug now.)

Reply:
the alternative is to fail to compile (`!Copy` types treat generic lengths as `> 1`, so if it later happens to be `1` that's irrelevant). While we could permit that, we'd be adding a new source of post monomorphization errors.

Reply:
I mean I am fine with this as long as it is guaranteed that this cannot ICE due to the safety net `delay_span_bug` this PR also adds.
I want to be sure that we do not draw ourselves into a corner where, or backwards compatibility reasons, we have to remove that safety net to fix an ICE.